src/gs-background-tile.c

```diff
@@ -0,0 +1,281 @@
+/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
+ *
+ * Copyright (C) 2013 Matthias Clasen <mclasen@redhat.com>
+ * Copyright (C) 2017 Joaquim Rocha <jrocha@endlessm.com>
+ *
+ * Licensed under the GNU General Public License Version 2
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "config.h"
+
+#include <glib/gi18n.h>
+#include <gtk/gtk.h>
+
+#include "gs-background-tile.h"
+#include "gs-star-widget.h"
+#include "gs-common.h"
+
+struct _GsBackgroundTile
+{
+	GsAppTile	 parent_instance;
+
+	GsApp		*app;
+	GtkWidget	*name_label;
+	GtkWidget	*summary_label;
+	GtkWidget	*image;
+	GtkWidget	*image_box;
+	GtkWidget	*installed_icon;
+	GtkWidget	*price;
+	GtkWidget	*stack;
+	gboolean	 key_colors_set;
+};
+
+G_DEFINE_TYPE (GsBackgroundTile, gs_background_tile, GS_TYPE_APP_TILE)
+
+static GsApp *
+gs_background_tile_get_app (GsAppTile *tile)
+{
+	return GS_BACKGROUND_TILE (tile)->app;
+}
+
+static void
+gs_background_tile_setup_colors_bg (GsBackgroundTile *tile)
+{
+	GsApp *app = tile->app;
+	const guint num_colors = 4;
+	const GPtrArray *colors = gs_app_get_key_colors (app);
+	/* initialise with 1 extra member because we need the last element to be NULL*/
+	g_auto(GStrv) gradients = g_new0 (gchar *, num_colors + 1);
+	g_autofree gchar *css = NULL;
+	g_autofree gchar *gradients_str = NULL;
+	gint degrees = 45;
+
+	if (colors->len == 0 || tile->key_colors_set)
```

Comment:
Sorry for the back and forth on this implementation, but this does not look quite right yet.

Right now `tile->key_colors_set` is only reset to `FALSE` when setting a new app on the tile, through `gs_background_tile_set_app()`, and set to `TRUE` at the end of this function. That means that any changes to the app's newly introduced `key-colors` property will effectively be ignored.

The code that looks at both the key colors and css and decides which one to set (right now part of `update_tile_info()`) should run when either of those notify a change.

Reply:
This was because the notify code can also be triggered when the name, summary, and more importantly, the state changes. And we don't want to recalculate the gradients whenever all these change. That's why I was making sure that the gradients were only set once...

The key-colors are being set from the apps' icons, and icons don't usually change. When they do, because an app may have completely changed, a refresh of the catalog in g-s will happen, so we'll rebuild the widget.

I can have a dedicated callback for _notify::key-colors_ if you prefer, and we'll just stop having this treated in the same function as all the other states. Do you prefer it this way?